backend/src/gamestate/game.rs

```diff
@@ -0,0 +1,97 @@
+use crate::cards::deck::Deck;
+use crate::gamestate::player::Player;
+use crate::gamestate::serialization::{FinishedStatus, LobbyStatus, RunningStatus};
+use crate::gamestate::WSMessage;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Clone, Copy)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum GameStatus {
+    Lobby,
+    Running,
+    Finished,
+}
+
+pub struct Game {
+    status: GameStatus,
+    pub players: Vec<Player>,
+    deck: Deck,
+    turns_played: usize,
+}
+
+impl Game {
+    pub fn new(author_name: String) -> Game {
+        Game {
+            status: GameStatus::Lobby,
+            players: vec![Player::new(author_name, true)],
+            deck: Deck::new(),
+            turns_played: 0,
+        }
+    }
+
+    pub fn find_player(&self, name: String) -> Option<&Player> {
+        self.players
+            .iter()
+            .find(|player| player.get_name_clone() == name)
+    }
+
+    pub fn find_author(&self) -> Option<&Player> {
+        self.players.iter().find(|player| player.is_author)
+    }
+
+    pub fn find_author_name(&self) -> String {
+        match self.find_author() {
+            None => "UnknownAuthor".into(),
+            Some(author) => author.get_name_clone(),
+        }
+    }
+
+    pub fn add_player(&mut self, name: String) {
+        self.players.push(Player::new(name, false))
+    }
+
+    pub fn get_finished_players(&self) -> Vec<&Player> {
+        let mut result = self
+            .players
+            .iter()
+            .filter(|p| p.is_finished())
+            .collect::<Vec<&Player>>();
+        result.sort_by_key(|player| player.get_position_unsafe());
+        result
+    }
+
+    pub fn get_finished_player_names(&self) -> Vec<String> {
+        self.get_finished_players()
+            .iter()
+            .map(|p| p.get_name_clone())
+            .collect()
+    }
```

Comment:
Je nějaký důvod, proč mapování nenechat na uživateli funkce `get_finished_players()`?

Reply:
Důvod byl, že se stejné mapování používá pro FinishedStatus i RunningStatus, tak jsem to extrahoval do jedné funkce.

Reply:
Ok, to beru, je to přece jen už delší kus kódu. Možná bych argumentoval tím, že je to specifické pro serializaci a tím pádem by to mělo být u ní, ale asi si tím nejsem natolik jistý, abych to protlačoval :D Resolved